### PR TITLE
10218 pre-commit autoupdate

### PR DESCRIPTION
**Pull request by graingert**
## Scope
update pre-commit hooks and exempt the pre-commit.ci autoupdate bot from newsfragments

## Contributor Checklist:

* [x] The associated ticket in Trac is here: https://twistedmatrix.com/trac/ticket/10218
* [ ] I ran `tox -e lint` to format my patch to meet the [Twisted Coding Standard](https://twistedmatrix.com/documents/current/core/development/policy/coding-standard.html)
* [x] I have created a newsfragment in src/twisted/newsfragments/ (see: [News files](https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles))
* [x] The title of the PR starts with the associated Trac ticket number (without the `#` character).
* [x] I have updated the automated tests and checked that all checks for the PR are green.
* [x] I have submitted the associated Trac ticket for review by adding the word `review` to the keywords field in Trac, and putting a link to this PR in the comment; it shows up in https://twisted.reviews/ now.
* [ ] The merge commit will use the below format
  The first line is automatically generated by GitHub based on PR ID and branch name.
  The other lines generated by GitHub should be replaced.

```
Merge pull request #123 from twisted/4356-branch-name-with-trac-id

Author: <github_username>, <github_usernames_if_more_authors>
Reviewer: <github_username>, <github_username_if_more_reviewers>
Fixes: ticket:<trac_ticket_number>, ticket:<another_if_more_in_one_PR>

Long description providing a summary of these changes.
(as long as you wish)
```

### CODE REVIEW

File: src/twisted/python/test/test_release.py [hunk 1]
```diff
@@ -1078,6 +1078,59 @@ def test_releaseWithNewsfragments(self):
...
+        with self.assertRaises(SystemExit) as e:
+            CheckNewsfragmentScript(logs.append).main([self.repo.path])
+
+        self.assertEqual(e.exception.args, (0,))
+        self.assertEqual(logs[-1], "Autoupdate branch with no newsfragments, all good.")
+
+    def test_preCommitAutoupdateWithNewsfragments(self):
+        """
+        Running it on a release branch returns red if there are new
+        newsfragments.
```
Comment (graingert) [inline src/twisted/python/test/test_release.py]:
```suggestion
        Running it on the autoupdate branch returns red if there are new
        newsfragments.
```

File: src/twisted/python/test/test_release.py [hunk 2]
```diff
@@ -1078,6 +1078,59 @@ def test_releaseWithNewsfragments(self):
         self.assertEqual(e.exception.args, (1,))
         self.assertEqual(logs[-1], "No newsfragments should be on the release branch.")
 
+    def test_preCommitAutoupdate(self):
+        """
+        Running it on an autoupdate branch returns green if there is no
```
Comment (graingert) [inline src/twisted/python/test/test_release.py]:
```suggestion
        Running it on the autoupdate branch returns green if there is no
```

File: src/twisted/python/_release.py [hunk 3]
```diff
@@ -581,6 +581,14 @@ def main(self, args):
                 self._print("Release branch with no newsfragments, all good.")
                 sys.exit(0)
 
+        if branch == "pre-commit-ci-update-config":
```
Comment (adiroiban) [inline src/twisted/python/_release.py]: I think that is best to use the "committer" .

The branch name is "mangled" by the GitHub pull_request 

```suggestion
        if os.environ.get("GITHUB_ACTOR", "") == "pre-commit-ci[bot]':
```

---------------

I started to update the script, but I didn't had enough time :(

To help with debugging, I was including the branch name in the error message.

In this way , is easier to observe if the branch name is not the expected one.

**Comment by adiroiban** [comment]
This is my diff

```diff
diff --git a/src/twisted/python/_release.py b/src/twisted/python/_release.py
index 6329e171f9..b687bb7bf2 100644
--- a/src/twisted/python/_release.py
+++ b/src/twisted/python/_release.py
@@ -582,12 +582,15 @@ class CheckNewsfragmentScript:
                 self._print("Release branch with no newsfragments, all good.")
                 sys.exit(0)
 
-        if branch == "pre-commit-ci-update-config":
+        if os.environ.get("GITHUB_ACTOR", "") == "pre-commit-ci[bot]":
+            # The run was triggered by pre-commit.ci.
             if newsfragments:
-                self._print("No newsfragments should be on the autoupdate branch.")
+                self._print(
+                    "No newsfragments should be present on an autoupdated branch."
+                )
                 sys.exit(1)
             else:
-                self._print("Autoupdate branch with no newsfragments, all good.")
+                self._print("Autoupdated branch with no newsfragments, all good.")
                 sys.exit(0)
 
         for change in newsfragments:
diff --git a/src/twisted/python/test/test_release.py b/src/twisted/python/test/test_release.py
index d751acdbf1..c71b0c1a49 100644
--- a/src/twisted/python/test/test_release.py
+++ b/src/twisted/python/test/test_release.py
@@ -53,6 +53,9 @@ if sys.platform != "win32":
 else:
     skip = "Release toolchain only supported on POSIX."
 
+# This should match the GitHub Actions environment used by pre-comnmit.ci to push changes to the auto-updated branches.
+PRECOMMIT_CI_ENVIRON = {"GITHUB_ACTOR": "pre-commit-ci[bot]"}
+
 
 class ExternalTempdirTestCase(TestCase):
     """
@@ -1094,12 +1097,15 @@ class CheckNewsfragmentScriptTests(ExternalTempdirTestCase):
         runCommand(["git", "commit", "-m", "some file"], cwd=self.repo.path)
 
         logs = []
+        self.patch(os, "environ", PRECOMMIT_CI_ENVIRON)
 
         with self.assertRaises(SystemExit) as e:
             CheckNewsfragmentScript(logs.append).main([self.repo.path])
 
         self.assertEqual(e.exception.args, (0,))
-        self.assertEqual(logs[-1], "Autoupdate branch with no newsfragments, all good.")
+        self.assertEqual(
+            logs[-1], "Autoupdated branch with no newsfragments, all good."
+        )
 
     def test_preCommitAutoupdateWithNewsfragments(self):
         """
@@ -1122,13 +1128,14 @@ class CheckNewsfragmentScriptTests(ExternalTempdirTestCase):
         runCommand(["git", "commit", "-m", "fragment"], cwd=self.repo.path)
 
         logs = []
+        self.patch(os, "environ", PRECOMMIT_CI_ENVIRON)
 
         with self.assertRaises(SystemExit) as e:
             CheckNewsfragmentScript(logs.append).main([self.repo.path])
 
         self.assertEqual(e.exception.args, (1,))
         self.assertEqual(
-            logs[-1], "No newsfragments should be on the autoupdate branch."
+            logs[-1], "No newsfragments should be present on an autoupdated branch."
         )
 
     def test_onlyQuotes(self):

```

**Comment by graingert** [comment]
@adiroiban your fix for pre-commit newsfragments exemption LGTM

**Review by adiroiban — APPROVED** [review]
LGTM!

**Pull request merged**